Add profile data listing and editing to occ command

salmart-dev opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

* Resolves: #49744

## Summary

This PR adds the ability to list, edit or delete profile properties through the `occ` command.

Before:

![image](https://github.com/user-attachments/assets/54a6edf2-d815-4528-bac6-04ed15dccc3b)

After:

![image](https://github.com/user-attachments/assets/0aea26a6-b83f-4fc5-8e75-ae2092f01bdf)

## TODO

- [x] Update tests once the solution is approved
- [ ] Open new issues for the encountered problems in the command
- [x] Check if documentation and/or wiki needs update
- [x] Verify correctness of `occ user:setting [uid] settings` and decide whether to open a new ticket or fix in this context.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

salmart-dev committed:
feat: allow reading profile properties from occ

This change adds support for reading profile information through the occ
command, and updates the corresponding test.

Signed-off-by: Salvatore Martire <4652631+klarkent@users.noreply.github.com>

provokateurin reviewed (commented):
I found the first commit to be really hard to review, but I think your refactoring is good and I trust you and the tests that the changes are correct :)

Could you please add some tests for the added functionality? Please also squash the last two commits so the changes are atomic, in case anyone in the future has to bisect the changes.

salmart-dev commented on tests/Core/Command/User/SettingTest.php:
```diff
@@ -16,6 +17,9 @@
 class SettingTest extends TestCase {
+
+	/** @var \OCP\Accounts\IAccountManager|\PHPUnit\Framework\MockObject\MockObject */
+	protected $accountManager;
```
Thanks, I followed your suggestion and applied the change to all the other properties. I also imported MockObject to avoid writing the fully qualified class name at each property, as it gets long and distracting.

provokateurin replied:
Perfect!

provokateurin commented:
> I'm sorry my first commit was hard to review. I wanted to keep the refactoring in one commit, but I see that it would've been better to split the changes more, at least for the sake of the review. If you have other suggestions for me about this, please let me know!

Like I said it's not that bad, especially given how bad the code was before. Sometimes rewriting is the only solution to make it clean.

> Regrding the fixes, do you prefer separate commits per item to be fixed, and then squash them with the relevant commits once the review is over, or something else?

Ideally you only do fixup commits, so the reviewers can easily check what you changed, and then squash once at the end when everything was approved and is ready to be merged.

> Also, while reading and implementing my changes, I noticed a few oddities and I'm not sure whether those should be discussed in this PR or in another place:

You could fix them here, but that might a bit too much. If you don't mind, please open some issues about your findings so we can fix them later.

salmart-dev commented on core/Command/User/Setting.php:
```diff
@@ -193,6 +206,20 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+	private function editProfileProperty($uid, $key, $value): void {
+		$user = $this->userManager->get($uid);
+		if (!$user) {
+			throw new \InvalidArgumentException('The user "' . $uid . '" must exist to edit this setting.');
```
While writing a test for it, I realised that this would just bubble the exception, while the usual pattern is to print an error message and return 1 as a return code for occ. Wouldn't it make sense to change this exception to follow that behaviour?

provokateurin replied:
Yes, either you log the error directly or keep throwing the exception and catch it later and then log it as an error.

salmart-dev replied:
I followed the current pattern.
I also noticed an error in my previous code where I passed the app name instead of the user id in a function call, fixed it and hardened the tests to check for this.

salmart-dev commented:
@provokateurin I finally managed to finish and push the updated/new tests 🎉

I decided to edit the get/list ones, since the test code would be not too complex, but added new versions for the set and delete ones since adding the new case to the old tests would make them a bit too complex and way harder to read.

Regarding my previous comment about the `--default-value` not working, I had a big 🤦 moment when I realised that it does exactly what it is supposed to and that I just completely missed its meaning (probably thanks to having fever).

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +146,143 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+		return 0;
+	}
+
+	/**
+	 * @param OutputInterface $output
+	 * @param $uid
+	 * @param $key
+	 * @return int
+	 * @throws PropertyDoesNotExistException if $key is not a property of the account.
+	 */
+	private function deleteProfileProperty(OutputInterface $output, $uid, $key): int {
```
```suggestion
	/**
	 * @throws PropertyDoesNotExistException if $key is not a property of the account.
	 */
	private function deleteProfileProperty(OutputInterface $output, string $uid, string $key): int {
```

salmart-dev replied:
It's fixed now

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +146,143 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+		return $this->editProfileProperty($output, $uid, $key, '');
+	}
+
+	/**
+	 * @param OutputInterface $output
+	 * @param $uid
+	 * @param $key
+	 * @param $value
+	 * @return int
+	 * @throws PropertyDoesNotExistException if $key is not a property of the account.
+	 */
+	private function editProfileProperty(OutputInterface $output, $uid, $key, $value): int {
```
```suggestion
	/**
	 * @throws PropertyDoesNotExistException if $key is not a property of the account.
	 */
	private function editProfileProperty(OutputInterface $output, string $uid, string $key, string $value): int {
```

salmart-dev replied:
It's fixed now

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +146,143 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+		}
+
+		return null;
+	}
+	private function setSettingsProperty(InputInterface $input, OutputInterface $output, string $uid, string $key): ?int {
```
```suggestion
	}

	private function setSettingsProperty(InputInterface $input, OutputInterface $output, string $uid, string $key): ?int {
```

salmart-dev replied:
It's fixed now.

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +146,143 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+		return null;
 	}
 
+	private function getStoredValue(mixed $uid, mixed $app, mixed $key): ?string {
```
```suggestion
	private function getStoredValue(string $uid, string $app, string $key): ?string {
```

salmart-dev replied:
I trusted PHPStorm's autocompletion too much here 😅 Ok, it's fixed!

provokateurin reviewed (changes requested):
Ideally you use strong types for all parameters and maybe you can also try setting `declare (strict_types=1);` and check if everything still works with that.

salmart-dev commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +146,143 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+		return 0;
+	}
+
+	private function deleteSettingsProperty(OutputInterface $output, $uid, $key): ?int {
```
It's fixed now

salmart-dev commented:
Hey @provokateurin thanks again for the loops of review!

I addressed everything up to your latest comments but I noticed that the fix to avoid returning the "settings" values is still in the branch. I want to confirm that I should remove it, correct? That would mean removing the if statement from the following code but keeping the assignment (Setting.php:302).

```php
if (!$app || $app === 'settings') {
	// Only add the display name if the user exists
	$settings['settings']['display_name'] = $user->getDisplayName();
}
`

salmart-dev commented:
> TBH I'm not knowledgeable about how it is supposed to work. I would think the check makes sense, but maybe you can elaborate a bit why you think it should be removed.

I would love to keep the fix, since it leads to a more correct output for the `occ` function, but I don't want to skip the process of opening a new ticket and fixing it in that context if not doing that would be a problem in any way. And since you mentioned earlier that fixing the issues I reported in this branch would be too much, I was asking if I should remove the fix or not.

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +148,135 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+	}
+
+	private function setSettingsProperty(InputInterface $input, OutputInterface $output, string $uid, string $key): ?int {
+		$user = $this->userManager->get($uid);
+		if ($user instanceof IUser) {
```
If the user does not exist you should log an error and return 1 instead of doing nothing.

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +148,135 @@ protected function execute(InputInterface $input, OutputInterface $output): int
-							return 1;
-						}
-					}
+			$this->config->setUserValue($uid, $app, $key, $input->getArgument('value'));
```
So if it's neither a profile nor a settings value, you set it as a normal user value? I'd strongly prefer if you put it in an else block, as currently it would also be executed when `setSettingsProperty` returns null (but I would like to have that changed as well as per my comment above).

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +148,135 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+	}
+
+	private function deleteSettingsProperty(OutputInterface $output, string $uid, string $key): ?int {
+		$user = $this->userManager->get($uid);
+		if ($user instanceof IUser) {
```
Here it should also be an error and never return null.

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +148,135 @@ protected function execute(InputInterface $input, OutputInterface $output): int
-				return 0;
-			} elseif ($input->hasParameterOption('--default-value')) {
-				$output->writeln($input->getOption('default-value'));
+			$this->config->deleteUserValue($uid, $app, $key);
```
Same here as above.

provokateurin commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +148,135 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+		$user = $this->userManager->get($uid);
+		if ($user instanceof IUser) {
+			if ($key === 'email') {
+				$user->setEMailAddress($input->getArgument('value'));
+			} elseif ($key === 'display_name') {
+				if (!$user->setDisplayName($input->getArgument('value'))) {
+					if ($user->getDisplayName() === $input->getArgument('value')) {
+						$output->writeln('<error>New and old display name are the same</error>');
+					} elseif ($input->getArgument('value') === '') {
```
The value should be passed to the method directly.

salmart-dev commented:
@provokateurin I adapted the `getStoredValue`, but I also had to add some handling to prevent reading, editing and deleting of properties that overlap between the "settings" and "profile" apps 🫠

come-nc commented on tests/Core/Command/User/SettingTest.php:
```diff
@@ -8,28 +8,33 @@
 	protected function setUp(): void {
 		parent::setUp();
 
+		$this->accountManager = $this->getMockBuilder(IAccountManager::class)
+			->disableOriginalConstructor()
+			->getMock();
```
```suggestion
		$this->accountManager = $this->createMock(IAccountManager::class);
```
Should be enough.

salmart-dev replied:
Addressed this in the new test file

come-nc reviewed (commented):
I fear that you’re gonna break `ignore-missing-user` option. You test in several places if the user exists despite the option value.
Also in some places you use the user object from the userManager without testing its value first.
And the test on $user is not always the same, it’s inconsistent, you have `if (!$user)` on one occasion and instanceof tests in the others.

It would be cleaner to get the user object early and pass it around, with a clear behavior when it’s `null`.
I was not able to test the branch right now, but I’ll try to do that later.

salmart-dev commented on core/Command/User/Setting.php:
```diff
@@ -143,82 +148,140 @@ protected function execute(InputInterface $input, OutputInterface $output): int
-				}
+		// prevent reading or editing of "profile" app properties exposed via the "settings" app
+		if ($this->isOverlappingSettingProperty($app, $key)) {
+			$output->writeln("<error>Setting '{$key}' belongs to the settings app.</error>");
```
This would prevent someone from running `occ user:setting alice profile email alice@gmx.de` by telling them that  `Setting 'email' belongs to the settings app.`, so they know they should run `occ user:setting alice settings email alice@gmx.de` instead.

salmart-dev commented:
> I fear that you’re gonna break `ignore-missing-user` option. You test in several places if the user exists despite the option value. Also in some places you use the user object from the userManager without testing its value first. And the test on $user is not always the same, it’s inconsistent, you have `if (!$user)` on one occasion and instanceof tests in the others.
> 
> It would be cleaner to get the user object early and pass it around, with a clear behavior when it’s `null`. I was not able to test the branch right now, but I’ll try to do that later.

What would be the expected behaviour when using `--ignore-missing-user` in the case of `profile` properties?
Setting the email, display_name and any property that can be set on the profile view of a user needs a user account from my understanding, so they cannot be added or edited if the user doesn't exist.

For example, in the implementation that existed for `email` and `display_name` if the user doesn't exist, instead of using the custom handling, the code inserts data using the `IConfig` interface which from my limited understanding would be wrong behaviour.

salmart-dev committed:
style(occ): clean up SettingTest class

Signed-off-by: Salvatore Martire <4652631+salmart-dev@users.noreply.github.com>

salmart-dev commented:
@provokateurin and @come-nc updates:

- rebased the branch on top of master
- undid the changes in Setting and SettingTest
- moved the new logic and tests to Profile and ProfileTest

The fixes for the issues with `occ user:setting setting` have also been undone. To avoid more churning of this PR I would avoid adding them back.

come-nc commented on core/Command/User/Profile.php:
```diff
@@ -0,0 +1,230 @@
+		// normalize uid
+		$input->setArgument('uid', $user->getUID());
+
+		if ($input->getArgument('key') === '' && $input->hasParameterOption('--default-value')) {
```
All these tests are testing key against `''` but it might be `null` if not passed I think.
I think the cleanest is something like:
```php
$key = (string)$input->getArgument('key');
if ($key === '') {
    /* All the tests for this situation */

salmart-dev replied:
The default value for key is `''` so the string casting should be unnecessary. I grouped the tests that check for `$key === ''` though.

come-nc commented on core/Command/User/Profile.php:
```diff
@@ -0,0 +1,230 @@
+			$this->checkInput($input);
+		} catch (\InvalidArgumentException $e) {
+			$output->writeln('<error>' . $e->getMessage() . '</error>');
+			return 1;
```
```suggestion
			return self::FAILURE;
```

come-nc commented on core/Command/User/Profile.php:
```diff
@@ -0,0 +1,230 @@
+		}
+
+		$uid = $input->getArgument('uid');
+		$key = $input->getArgument('key');
```
```suggestion
		$key = (string)$input->getArgument('key');
```

salmart-dev replied:
This should not be necessary as the default value for the `key` argument is `''`. Do we need to cover for `null` anyways?

come-nc commented on core/Command/User/Profile.php:
```diff
@@ -0,0 +1,230 @@
+
+		$uid = $input->getArgument('uid');
+		$key = $input->getArgument('key');
+		$user = $this->userManager->get($uid);
```
It would be better to somehow carry along the user object fetched in checkInput, because here technically it could be null (corner case like deleted user by another request).

salmart-dev replied:
I would argue that if we want to cover for the case where the user is deleted in a parallel request then we should never pass the user around, since using a cached user would either cause errors because of the user not existing any longer or worse, depending on how hardened the functions related to editing users are, re-add data into the user table for a user that has just been deleted.

come-nc commented on core/Command/User/Profile.php:
```diff
@@ -0,0 +1,230 @@
+		if ($key === '') {
+			$settings = $this->getAllProfileProperties($userAccount);
+			$this->writeArrayInOutputFormat($input, $output, $settings);
+			return 0;
```
```suggestion
			return self::SUCCESS;
```
Same change on all returns

come-nc commented on core/Command/User/Profile.php:
```diff
@@ -0,0 +1,234 @@
+	 * @param CompletionContext $context
+	 * @return string[]
+	 */
+	public function completeArgumentValues($argumentName, CompletionContext $context) {
```
```suggestion
	public function completeArgumentValues($argumentName, CompletionContext $context): array {
```

salmart-dev replied:
Done too

come-nc commented on tests/Core/Command/User/ProfileTest.php:
```diff
@@ -0,0 +1,489 @@
+
+	/**
+	 * @dataProvider dataCheckInput
+	 *
+	 * @param array $arguments
+	 * @param array $options
+	 * @param array $parameterOptions
+	 * @param bool $existingUser
+	 * @param string|null $expectedException
```
```suggestion
```
avoid adding phpdoc typing unless you add descriptions in there, otherwise it’s just duplicated information from the parameter list.
It only makes sense if the phpdoc type is more precise, like list<string> for an array.

st3iny reviewed (commented):
There are a lot of unrelated autoloader changes. Please revert them.

st3iny reviewed (commented):
For example this one

st3iny commented on apps/admin_audit/composer/autoload.php:
```diff
@@ -14,10 +14,7 @@
-        $err,
-        E_USER_ERROR
-    );
+    throw new RuntimeException($err);
```
For example this one

st3iny replied:
Exactly. Generally, only those two files need to be checked in.

st3iny replied:
PS: The differences usually come from differing versions of composer.

salmart-dev committed:
chore: update autoloader

Signed-off-by: Salvatore Martire <4652631+salmart-dev@users.noreply.github.com>

st3iny reviewed (approved):
Tested and works

salmart-dev commented:
Merged in https://github.com/nextcloud/server/pull/53540
